### PR TITLE
[14.0][ADD] account_reconciliation_widget

### PR DESCRIPTION
**Pull request by ozono**
First working alpha version.

Please review and send your comments.

Known issues:

- Some dirty code.
- JS tests not yet reviewed and restored.
- Python tests works, but would be convenient to write more.

### CODE REVIEW

**Comment by pedrobaeza** [comment]
Works OK on a first functional test, although it's very weird all the suspense account + post thing.

File: account_reconciliation_widget/models/account_journal.py [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+
+    def action_open_reconcile_to_check(self):
+        self.ensure_one()
+        ids = self.to_check_ids().ids
+        action_context = {
+            "show_mode_selector": False,
+            "company_ids": self.mapped("company_id").ids,
+        }
+        action_context.update({"suspense_moves_mode": True})
+        action_context.update({"statement_line_ids": ids})
```
Comment (PieterPaulussen) [inline account_reconciliation_widget/models/account_journal.py]: Why use `update` if you are instantiating the dictionary 2 lines up?
```
        action_context = {
            "show_mode_selector": False,
            "company_ids": self.mapped("company_id").ids,
            "suspense_moves_mode": True,
            "statement_line_ids": ids,
        }
```

File: account_reconciliation_widget/models/account_move.py [hunk 2]
```diff
@@ -0,0 +1,127 @@
...
+    _inherit = "account.move.line"
+
+    def _create_writeoff(self, writeoff_vals):
+        """Create a writeoff move per journal for the account.move.lines in
+         self. If debit/credit is not specified in vals, the writeoff amount
+         will be computed as the sum of amount_residual of the given recordset.
+
+        :param writeoff_vals: list of dicts containing values suitable for
+            account_move_line.create(). The data in vals will be processed to
+            create bot writeoff account.move.line and their enclosing
```
Comment (PieterPaulussen) [inline account_reconciliation_widget/models/account_move.py]: `both`

File: account_reconciliation_widget/models/reconciliation_widget.py [hunk 3]
```diff
@@ -0,0 +1,1165 @@
...
+            "|",
+            ("move_id.name", "ilike", search_str),
+            "|",
+            ("move_id.ref", "ilike", search_str),
+            "|",
+            ("date_maturity", "like", parse_date(self.env, search_str)),
+            "&",
+            ("name", "!=", "/"),
+            ("name", "ilike", search_str),
+        ]
```
Comment (PieterPaulussen) [inline account_reconciliation_widget/models/reconciliation_widget.py:669]: I know Odoo never tackled these types of things but the use of `ilike` can give significant performance hits on large tables. Even more `ilike`s in the same domain will probably never even return a result.

File: account_reconciliation_widget/models/account_journal.py (hunk 1 above)
Comment (ozono) [reply account_reconciliation_widget/models/account_journal.py]: You're right.

File: account_reconciliation_widget/models/account_move.py (hunk 2 above)
Comment (ozono) [reply account_reconciliation_widget/models/account_move.py]: Ok.

File: account_reconciliation_widget/models/reconciliation_widget.py (hunk 3 above)
Comment (ozono) [reply account_reconciliation_widget/models/reconciliation_widget.py:669]: I agree about this domain performance... but here it returns desired lines and has been a long time without reported issues. Also, this function can be easy overrided in a custom module.

**Review by pedrobaeza — APPROVED** [review]

**Comment by pedrobaeza** [comment]
Let's go with this as there are no more comments. Other later PRs can improve anything discovered in real tests:

/ocabot merge nobump

Thanks @ozono for the hardest work.

**Comment by randall-vx** [comment]
@hbto

**Comment by hbto** [comment]
@randall-vx ACK.

Thanks a lot.

Regards.

File: account_reconciliation_widget/models/reconciliation_widget.py (hunk 3 above)
Comment (gdgellatly) [reply account_reconciliation_widget/models/reconciliation_widget.py:669]: @PieterPaulussen This is not really true except for in a very basic untuned postgres installation. pg_trgm is a longstanding solution to openended ilike searches and any integrator who deals in mid sized tables of a few hundred thousand to a few tens of millions will already be monitoring for these and adding appropriate indexes.

The killer in that query with large tables will be more likely the generated IN clauses in non specific searches.

Comment (PieterPaulussen) [reply account_reconciliation_widget/models/reconciliation_widget.py:669]: A trigram index is indeed a good solution in some cases. Unfortunately just adding indexes won't suffice when `account.move.line` grows even further to roughly 2.5 billion records as we have in of our long running projects.
We solved the 'where id in []' issue a couple of years ago by changing the clause to use a custom stored boolean flag. However, even with that implementation, it was insufficient.
If I remember correctly it was the join operations from AML with `account.move` and `account.account` which were even worse.
In the end we replaced the search box with a more elegant solution allowing users to search specific fields with different operators according to their needs.

Comment (gdgellatly) [reply account_reconciliation_widget/models/reconciliation_widget.py:669]: @PieterPaulussen OK but that is not typical of any Odoo installation. Extreme outliers in terms of pushing Odoo have always had to rewrite those parts themself. To have that many move lines you have already done extensive customizations, changing the column type of id field, support for either BIGSERIAL or GUID style identifiers. Realistically, to have over 2.5bn records of just the small band of narrow accounts this is pre filtered to, that is way outside normal use.

Comment (PieterPaulussen) [reply account_reconciliation_widget/models/reconciliation_widget.py:669]: Agreed, but I found it at least noteworthy to mention the issue on the PR.
If Odoo keeps growing at their current rate, its only a matter of time before supporting large datasets out of the box becomes mandatory. For OCA modules, should we not try to take it into account as much as possible when we can clearly foresee these types of issues? Solving this uniquely for each customer puts a heavy strain on project budgets, and might push customers away from the ERP.
Thanks for your comments though, you do make a good point.

Comment (pedrobaeza) [reply account_reconciliation_widget/models/reconciliation_widget.py:669]: If you have any suggestion for solving the performance bottleneck without losing features, please propose it as a new PR.

**Comment by alexis-via** [comment]
Thanks for this big work on account_reconcile_widget!
I started to use it and I'm facing a problem: in the reconcile interface, I don't see the 'name' of the bank statement line (which is the same as the 'name' of the account.move.line) ; I only see the account move number... so the accountant is missing a very important information to work on the line!

bank_reconcile_label_issue

**Comment by alexis-via** [comment]
I'm facing another issue: as you can see in the screenshot above, I have a button "Frais bancaires" (account.reconcile.model). This button has type = 'writeoff_button' and 1 account.reconcile.model.line configured as 100% on an expense account. When I click on that button, nothing happens. Do you face the same pb ?

**Comment by alexis-via** [comment]
Remember that the field 'name' of account.bank.statement.line became 'payment_ref' in odoo v14. This is certainly the cause of the first issue I reported.

**Comment by alexis-via** [comment]
Before:

before

After: you have the proper label.

**Comment by moaadbourhim** [comment]
> I'm facing another issue: as you can see in the screenshot above, I have a button "Frais bancaires" (account.reconcile.model). This button has type = 'writeoff_button' and 1 account.reconcile.model.line configured as 100% on an expense account. When I click on that button, nothing happens. Do you face the same pb ?

@alexis-via  Did you find a solution for this problem ?

**Comment by alexis-via** [comment]
@moaadbourhim No, because I haven't tried to fix it so far. But we really need to fix it now.

**Comment by pedrobaeza** [comment]
#372 fixes the behavior with reconciliation models of type `writeoff_button`.

**Comment by moaadbourhim** [comment]
@pedrobaeza  @alexis-via 
I think there one more problem, when you choose for example one analytic Account from the list (Analytic Acc.) and then when you check the cost and Revenue of this Analytic account you find the duplication of the lines in "account.analytic.line".

You can see that in the images 

Choose for example Asustek - Wood Corner in analytic account :

analytic_acc


Check the Cost and Revenue of this Analytic Account :

cost_revenue


The list of cost and Revenue :

res

**Pull request merged**